README.md
# Lovelace animated weather card

Improved version of [weather card](https://github.com/bramkragten/weather-card) to support new hourly forecasts in Home Assistant 2023 and forward. Also inspired by the fork [Carte Lovelace Météo France](https://github.com/hacf-fr/lovelace-meteofrance-weather-card).

This card uses the awesome [animated SVG weather icons by amCharts](https://www.amcharts.com/free-animated-svg-weather-icons/).

![Weather Card](https://github.com/bramkragten/custom-ui/blob/master/weather-card/weather-card.gif?raw=true)

## Installation:

You have 2 main options, HACS or manual.

# HACS (recommended):

Install this repo as a custom repository in [HACS](https://hacs.xyz/): https://github.com/SweVictor/v-weather-card.

# Manual:

Download the contents of the `dist` folder to `/config/www/community/v-weather-card/`. (or an other folder in `/config/www/`)

# Add resources

Add the following to resources in your lovelace config (typically `/config/lovelace.yaml`):

```yaml
resources:
  - url: /hacsfiles/v-weather-card/v-weather-card.js
    type: module
```

## Configuration:

And add a card with type `custom:v-weather-card`:

```yaml
type: custom:v:weather-card
entity: weather.yourweatherentity
name: Optional name
```

If you want to use your local icons add the location to the icons:

```yaml
type: custom:v-weather-card
entity: weather.yourweatherentity
icons: "/local/custom-components/v-weather-card/icons/"
```

You can choose wich elements of the weather card you want to show:

The 3 different rows, being:

- The current weather icon, the current temperature and title
- The details about the current weather
- The X day forecast or hourly forecast

```yaml
type: custom:v-weather-card
entity: weather.yourweatherentity
current: true
details: false
forecast: true
hourly_forecast: false
number_of_forecasts: 5
```

If you want to show the sunrise and sunset times, make sure the `sun` component is enabled:

```yaml
# Example configuration.yaml entry
sun:
```

### Dark Sky:

When using Dark Sky you should put the mode to `daily` if you want a daily forecast with highs and lows.

```yaml
# Example configuration.yaml entry
weather:
  - platform: darksky
    api_key: YOUR_API_KEY
    mode: daily
```

### OpenWeather Map:

When using OpenWeather map you can select hourly(default) or daily forecast to show.

```yaml
# Example configuration.yaml entry
weather:
  - platform: openweathermap
    api_key: YOUR_API_KEY
```


CHANGELOG.md
## 1.0-beta

- First version of v-weather icons based on (weather-card)[https://github.com/bramkragten/weather-card]


src/card.ts
import { html, LitElement, TemplateResult, nothing, PropertyValues } from "lit";
import { styles } from "./card.styles";
import { customElement, state } from "lit/decorators";

import { HassEntity } from "home-assistant-js-websocket";
import { ActionConfig, ActionHandlerEvent, formatTime, HomeAssistant, LovelaceCardConfig, handleAction, hasAction, hasConfigOrEntityChanged } from "custom-card-helpers";

interface Config extends LovelaceCardConfig {
  header?: string;
  entity: string;

  type: string;
  name?: string;
  // show_warning?: boolean;
  // show_error?: boolean;
  // test_gui?: boolean;
  // entity?: string;
  tap_action?: ActionConfig;
  hold_action?: ActionConfig;
  double_tap_action?: ActionConfig;
}

const weatherIconsDay = {
  clear: "day",
  "clear-night": "night",
  cloudy: "cloudy",
  fog: "cloudy",
  hail: "rainy-7",
  lightning: "thunder",
  "lightning-rainy": "thunder",
  partlycloudy: "cloudy-day-3",
  pouring: "rainy-6",
  rainy: "rainy-5",
  snowy: "snowy-6",
  "snowy-rainy": "rainy-7",
  sunny: "day",
  windy: "cloudy",
  "windy-variant": "cloudy-day-3",
  exceptional: "!!",
};

const weatherIconsNight = {
  ...weatherIconsDay,
  clear: "night",
  sunny: "night",
  partlycloudy: "cloudy-night-3",
  "windy-variant": "cloudy-night-3",
};

const windDirections = [
  "N",
  "NNE",
  "NE",
  "ENE",
  "E",
  "ESE",
  "SE",
  "SSE",
  "S",
  "SSW",
  "SW",
  "WSW",
  "W",
  "WNW",
  "NW",
  "NNW",
  "N",
];

@customElement('v-weather-card')
export class VWeatherCard extends LitElement {
  // internal reactive states
  @state() private config!: Config;
  @state() private header: string | typeof nothing;
  @state() private entity: string;
  @state() private name: string;

  // @state() private _name: string;
  @state() private state: HassEntity;
  @state() private status: string;

  private numberElements = 0;

  private _hass: HomeAssistant;

  // lifecycle interface
  setConfig(config: Config) {
    // The config object contains the configuration specified by the user in ui-lovelace.yaml

    // It will minimally contain:
    // config.type = "custom:my-custom-card"

    // setConfig will ALWAYS be called at the start of the lifetime of the card
    // BEFORE the `hass` object is first provided.
    // It MAY be called several times during the lifetime of the card, e.g. if the configuration
    // of the card is changed.

    if (!config.entity) {
      // If no entity was specified, this will display a red error card with the message below
      throw new Error('You need to define an entity');
    }

    this.config = config;
  }

  getCardSize() {
    // The height of the card. 
    // Home Assistant uses this to automatically distribute all cards over the available columns.
    return 1;
  }

  set hass(hass: HomeAssistant) {
    this._hass = hass;
    console.debug("Setting hass", hass);
    this.state = hass.states[this.config.entity];
    console.debug("State", this.state);
    this.entity = this.config.entity;
    this.header = this.config.header === "" ? nothing : this.config.header;

    if (this.state) {
      this.status = this.state.state;
      let fn = this.state.attributes.friendly_name;
      this.name = fn ? fn : this.config.entity;
    }
    console.debug("Entity", this.entity);
    // console.log("Header", this.header);
    // console.log("Name", this.name);
    // console.log("State", this.state);
    console.debug("Status", this.status);

  //   this._hass.callApi("POST", "services/weather/get_forecast", {
  //     target: {
  //       entity_id: this.entity,
  //     },
  //     data: {
  //       type: "hourly"
  //     },
  //     response_variable: "weather_forecast"
  //   }).then(r => {
  //     console.log("Response: ", r);
  //   })
  }

  
  // declarative part
  static styles = styles;

  // https://lit.dev/docs/components/lifecycle/#reactive-update-cycle-performing
  protected shouldUpdate(changedProps: PropertyValues): boolean {
    if (!this.config) {
      return false;
    }

    return hasConfigOrEntityChanged(this, changedProps, false);
  }



  // https://lit.dev/docs/components/rendering/
  render(): TemplateResult {

    if (!this.config || !this._hass) {
      return html``;
    }


    if (this.config.hou)

    this.numberElements = 0;

    const lang = this._hass.selectedLanguage || this._hass.language;
    const stateObj = this._hass.states[this.config.entity];

    if (!stateObj) {
      return html`
        <style>
          .not-found {
            flex: 1;
            background-color: yellow;
            padding: 8px;
          }
        </style>
        <ha-card>
          <div class="not-found">
            Entity not available: ${this.config.entity}
          </div>
        </ha-card>
      `;
    }

    return html`
      <ha-card header="${this.header}" @click="${this._handleClick}">
        ${this.config.current !== false ? this.renderCurrent(stateObj) : ""}
        ${this.config.details !== false ? this.renderDetails(stateObj, this._hass.locale) : ""}
        ${this.config.forecast !== false
        ? this.renderForecast(stateObj, this._hass.locale)
        : ""}
      </ha-card>
      `;

  }

  fireEvent(node, type, detail, options: any = {}) {
    options = options || {};
    detail = detail === null || detail === undefined ? {} : detail;
    const event = new Event(type, {
      bubbles: options.bubbles === undefined ? true : options.bubbles,
      cancelable: Boolean(options.cancelable),
      composed: options.composed === undefined ? true : options.composed,
    });
    (event as any).detail = detail;
    node.dispatchEvent(event);
    return event;
  };

  _handleClick() {
    this.fireEvent(this, "hass-more-info", { entityId: this.config.entity });
  }

  private _handleAction(ev: ActionHandlerEvent): void {
    if (this._hass && this.config && ev.detail.action) {
      handleAction(this, this._hass, this.config, ev.detail.action);
    }
  }



  renderCurrent(stateObj) {
    this.numberElements++;

    return html`
      <div class="current ${this.numberElements > 1 ? "spacer" : ""}">
        <span
          class="icon bigger"
          style="background: none, url('${this.getWeatherIcon(
      stateObj.state.toLowerCase(),
      this._hass.states["sun.sun"]
    )}') no-repeat; background-size: contain;"
          >${stateObj.state}
        </span>
        ${this.config.name
        ? html` <span class="title"> ${this.config.name} </span> `
        : ""}
        <span class="temp"
          >${stateObj.attributes.temperature_unit === "°F"
        ? Math.round(stateObj.attributes.temperature)
        : stateObj.attributes.temperature}</span
        >
        <span class="tempc">
          ${stateObj.attributes.temperature_unit}
        </span>
      </div>
    `;
  }

  renderDetails(stateObj, userLocale) {
    const sun = this._hass.states["sun.sun"];
    let next_rising;
    let next_setting;

    if (sun) {
      next_rising = formatTime(new Date(sun.attributes.next_rising), userLocale);
      next_setting = formatTime(new Date(sun.attributes.next_setting), userLocale);
    }

    this.numberElements++;

    return html`
      <ul class="variations ${this.numberElements > 1 ? "spacer" : ""}">
        <li>
          <ha-icon icon="mdi:water-percent"></ha-icon>
          ${stateObj.attributes.humidity}<span class="unit"> % </span>
        </li>
        <li>
          <ha-icon icon="mdi:weather-windy"></ha-icon> ${windDirections[
      Math.floor(((stateObj.attributes.wind_bearing as number) + 11.25) / 22.5)
      ]}
          ${stateObj.attributes.wind_speed}<span class="unit">
            ${stateObj.attributes.wind_speed_unit}
          </span>
        </li>
        <li>
          <ha-icon icon="mdi:gauge"></ha-icon>
          ${stateObj.attributes.pressure}
          <span class="unit">
            ${stateObj.attributes.pressure_unit}
          </span>
        </li>
        ${stateObj.attributes.visibility && html`
        <li >
          <ha-icon icon="mdi:weather-fog"></ha-icon> ${stateObj.attributes
          .visibility}<span class="unit">
            ${stateObj.attributes.visibility_unit}
          </span>
        </li>
        `}
        ${next_rising
        ? html`
              <li>
                <ha-icon icon="mdi:weather-sunset-up"></ha-icon>
                ${next_rising}
              </li>
            `
        : ""}
        ${next_setting
        ? html`
              <li>
                <ha-icon icon="mdi:weather-sunset-down"></ha-icon>
                ${next_setting}
              </li>
            `
        : ""}
      </ul>
    `;
  }

  renderForecast(stateObj, userLocale) {
    let forecast = stateObj.attributes.forecast;
    
    if (this.config.hourly_forecast) {
      const sensorName = `sensor.${this.entity.replace('.','_')}_hourly`;
      // console.debug("Trickery with hourly forecast! ", sensorName);
      forecast = this._hass.states[sensorName].attributes.forecast;
    }


    if (!forecast || forecast.length === 0) {
      return html``;
    }

    this.numberElements++;
    return html`
      <div class="forecast clear ${this.numberElements > 1 ? "spacer" : ""}">
        ${forecast
        .slice(
          0,
          this.config.number_of_forecasts
            ? this.config.number_of_forecasts
            : 5
        )
        .map(
          (daily) => html`
              <div class="day">
                <div class="dayname">
                  ${this.config.hourly_forecast
              ? formatTime(new Date(daily.datetime),userLocale)
              : new Date(daily.datetime).toLocaleDateString(userLocale.language, {
                weekday: "short",
              })}
                </div>
                <i
                  class="icon"
                  style="background: none, url('${this.getWeatherIcon(
                daily.condition.toLowerCase()
              )}') no-repeat; background-size: contain"
                ></i>
                <div class="highTemp">
                  ${daily.temperature}
                  ${stateObj.attributes.temperature_unit}
                </div>
                ${daily.templow !== undefined
              ? html`
                      <div class="lowTemp">
                        ${daily.templow}
                        ${stateObj.attributes.temperature_unit}
                      </div>
                    `
              : ""}
                ${!this.config.hide_precipitation &&
              daily.precipitation !== undefined &&
              daily.precipitation !== null
              ? html`
                      <div class="precipitation">
                        ${Math.round(daily.precipitation * 10) / 10} 
                        ${stateObj.attributes.precipitation_unit}
                      </div>
                    `
              : ""}
                ${!this.config.hide_precipitation &&
              daily.precipitation_probability !== undefined &&
              daily.precipitation_probability !== null
              ? html`
                      <div class="precipitation_probability">
                        ${Math.round(daily.precipitation_probability)} 
                        ${stateObj.attributes.percipitation_unit}
                      </div>
                    `
              : ""}
              </div>
            `
        )}
      </div>
    `;
  }



  getWeatherIcon(condition, sun = null) {
    return `${this.config.icons
      ? this.config.icons
      : "https://cdn.jsdelivr.net/gh/bramkragten/weather-card/dist/icons/"
      }${sun && sun.state == "below_horizon"
        ? weatherIconsNight[condition]
        : weatherIconsDay[condition]
      }.svg`;
  }

  // card configuration
  static getConfigElement() {
    return document.createElement("toggle-card-typescript-editor");
  }

  static getStubConfig() {
    return {
      entity: "weather.home",
      header: "",
    };
  }
}


src/index.ts
import "./card";
import { ToggleCardTypeScriptEditor } from "./editor";

declare global {
  interface Window {
    customCards: Array<Object>;
  }
}

// customElements.define("v-weather-card", VWeatherCard);
customElements.define(
  "v-weather-card-editor",
  ToggleCardTypeScriptEditor
);

window.customCards = window.customCards || [];
window.customCards.push({
  type: "v-weather-card",
  name: "V Weather Card",
  description: "Animated weather card for Home Assistant",
});


// // This puts your card into the UI card picker dialog
// (window as any).customCards = (window as any).customCards || [];
// (window as any).customCards.push({
//   type: 'v-weather-card',
//   name: 'V Weather Card',
//   description: 'A custom weather card with animated icons',
//   preview: true,
//   documentationURL: "https://github.com/SweVictor/v-weather-card",
// });


src/card.styles.ts
import { css } from "lit";

export const styles = css`
    ha-card {
      cursor: pointer;
      margin: auto;
      overflow: hidden;
      padding-top: 1.3em;
      padding-bottom: 1.3em;
      padding-left: 1em;
      padding-right: 1em;
      position: relative;
    }

    .spacer {
      padding-top: 1em;
    }

    .clear {
      clear: both;
    }

    .title {
      position: absolute;
      left: 3em;
      font-weight: 300;
      font-size: 3em;
      color: var(--primary-text-color);
    }

    .temp {
      font-weight: 300;
      font-size: 4em;
      color: var(--primary-text-color);
      position: absolute;
      right: 1em;
    }

    .tempc {
      font-weight: 300;
      font-size: 1.5em;
      vertical-align: super;
      color: var(--primary-text-color);
      position: absolute;
      right: 1em;
      margin-top: -14px;
      margin-right: 7px;
    }

    @media (max-width: 460px) {
      .title {
        font-size: 2.2em;
        left: 4em;
      }
      .temp {
        font-size: 3em;
      }
      .tempc {
        font-size: 1em;
      }
    }

    .current {
      padding: 1.2em 0;
      margin-bottom: 3.5em;
    }

    .variations {
      display: flex;
      flex-flow: row wrap;
      justify-content: space-between;
      font-weight: 300;
      color: var(--primary-text-color);
      list-style: none;
      padding: 0 1em;
      margin: 0;
    }

    .variations ha-icon {
      height: 22px;
      margin-right: 5px;
      color: var(--paper-item-icon-color);
    }

    .variations li {
      flex-basis: auto;
      width: 50%;
    }

    .variations li:nth-child(2n) {
      text-align: right;
    }

    .variations li:nth-child(2n) ha-icon {
      margin-right: 0;
      margin-left: 8px;
      float: right;
    }

    .unit {
      font-size: 0.8em;
    }

    .forecast {
      width: 100%;
      margin: 0 auto;
      display: flex;
    }

    .day {
      flex: 1;
      display: block;
      text-align: center;
      color: var(--primary-text-color);
      border-right: 0.1em solid #d9d9d9;
      line-height: 2;
      box-sizing: border-box;
    }

    .dayname {
      text-transform: uppercase;
    }

    .forecast .day:first-child {
      margin-left: 0;
    }

    .forecast .day:nth-last-child(1) {
      border-right: none;
      margin-right: 0;
    }

    .highTemp {
      font-weight: bold;
    }

    .lowTemp {
      color: var(--secondary-text-color);
    }

    .precipitation {
      color: var(--primary-text-color);
      font-weight: 300;
    }

    .icon.bigger {
      width: 10em;
      height: 10em;
      margin-top: -4em;
      position: absolute;
      left: 0em;
    }

    .icon {
      width: 50px;
      height: 50px;
      margin-right: 5px;
      display: inline-block;
      vertical-align: middle;
      background-size: contain;
      background-position: center center;
      background-repeat: no-repeat;
      text-indent: -9999px;
    }

    .weather {
      font-weight: 300;
      font-size: 1.5em;
      color: var(--primary-text-color);
      text-align: left;
      position: absolute;
      top: -0.5em;
      left: 6em;
      word-wrap: break-word;
      width: 30%;
    }
  `;


src/v-weather-card.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import { LitElement, html, TemplateResult, css, PropertyValues, CSSResultGroup } from 'lit';
import { customElement, property, state } from 'lit/decorators';
import {
  HomeAssistant,
  hasConfigOrEntityChanged,
  hasAction,
  ActionHandlerEvent,
  handleAction,
  LovelaceCardEditor,
  getLovelace,
} from 'custom-card-helpers'; // This is a community maintained npm module with common helper functions/types. https://github.com/custom-cards/custom-card-helpers

import type { BoilerplateCardConfig } from './types';

import { localize } from './localize/localize';


// const LitElement = customElements.get("ha-panel-lovelace") ? Object.getPrototypeOf(customElements.get("ha-panel-lovelace")) : Object.getPrototypeOf(customElements.get("hc-lovelace"));
// const html = LitElement.prototype.html;
// const css = LitElement.prototype.css;

const weatherIconsDay = {
  clear: "day",
  "clear-night": "night",
  cloudy: "cloudy",
  fog: "cloudy",
  hail: "rainy-7",
  lightning: "thunder",
  "lightning-rainy": "thunder",
  partlycloudy: "cloudy-day-3",
  pouring: "rainy-6",
  rainy: "rainy-5",
  snowy: "snowy-6",
  "snowy-rainy": "rainy-7",
  sunny: "day",
  windy: "cloudy",
  "windy-variant": "cloudy-day-3",
  exceptional: "!!",
};

const weatherIconsNight = {
  ...weatherIconsDay,
  clear: "night",
  sunny: "night",
  partlycloudy: "cloudy-night-3",
  "windy-variant": "cloudy-night-3",
};

const windDirections = [
  "N",
  "NNE",
  "NE",
  "ENE",
  "E",
  "ESE",
  "SE",
  "SSE",
  "S",
  "SSW",
  "SW",
  "WSW",
  "W",
  "WNW",
  "NW",
  "NNW",
  "N",
];

// This puts your card into the UI card picker dialog
(window as any).customCards = (window as any).customCards || [];
(window as any).customCards.push({
  type: 'v-weather-card',
  name: 'V Weather Card',
  description: 'A custom weather card with animated icons',
  preview: true,
  documentationURL: "https://github.com/SweVictor/v-weather-card",
});


const fireEvent = (node, type, detail, options) => {
  options = options || {};
  detail = detail === null || detail === undefined ? {} : detail;
  const event = new Event(type, {
    bubbles: options.bubbles === undefined ? true : options.bubbles,
    cancelable: Boolean(options.cancelable),
    composed: options.composed === undefined ? true : options.composed,
  });
  (event as any).detail = detail;
  node.dispatchEvent(event);
  return event;
};

// function hasConfigOrEntityChanged(element, changedProps) {
//   if (changedProps.has("_config")) {
//     return true;
//   }

//   const oldHass = changedProps.get("hass");
//   if (oldHass) {
//     return (
//       oldHass.states[element._config.entity] !==
//         element.hass.states[element._config.entity] ||
//       oldHass.states["sun.sun"] !== element.hass.states["sun.sun"]
//     );
//   }

//   return true;
// }


@customElement('v-weather-card')
export class VWeatherCard extends LitElement {
  public static async getConfigElement(): Promise<LovelaceCardEditor> {
    await import('./v-weather-card-editor.js');
    return document.createElement("v-weather-card-editor");
  }

  public static getStubConfig(): Record<string, unknown> {
    return {};
  }
  // static getStubConfig(hass, unusedEntities, allEntities) {
  //   let entity = unusedEntities.find((eid) => eid.split(".")[0] === "weather");
  //   if (!entity) {
  //     entity = allEntities.find((eid) => eid.split(".")[0] === "weather");
  //   }
  //   return { entity };
  // }

  // TODO Add any properities that should cause your element to re-render here
  // https://lit.dev/docs/components/properties/
  @property({ attribute: false }) public hass!: HomeAssistant;

  @state() private config!: BoilerplateCardConfig;

  // https://lit.dev/docs/components/properties/#accessors-custom
  public setConfig(config: BoilerplateCardConfig): void {
    // TODO Check for required fields and that they are of the proper format
    if (!config) {
      throw new Error(localize('common.invalid_configuration'));
    }

    if (config.test_gui) {
      getLovelace().setEditMode(true);
    }

    this.config = {
      name: 'Boilerplate',
      ...config,
    };
  }

  // https://lit.dev/docs/components/lifecycle/#reactive-update-cycle-performing
  protected shouldUpdate(changedProps: PropertyValues): boolean {
    if (!this.config) {
      return false;
    }

    return hasConfigOrEntityChanged(this, changedProps, false);
  }





  // static get properties() {
  //   return {
  //     _config: {},
  //     hass: {},
  //   };
  // }


  // https://lit.dev/docs/components/rendering/
  protected render(): TemplateResult | void {
    // TODO Check for stateObj or other necessary things and render a warning if missing
    if (this.config.show_warning) {
      return this._showWarning(localize('common.show_warning'));
    }

    if (this.config.show_error) {
      return this._showError(localize('common.show_error'));
    }

    return html`
      <ha-card
        .header=${this.config.name}
        @action=${this._handleAction}
        .actionHandler=${actionHandler({
          hasHold: hasAction(this.config.hold_action),
          hasDoubleClick: hasAction(this.config.double_tap_action),
        })}
        tabindex="0"
        .label=${`V Weather Card: ${this.config.entity || 'No Entity Defined'}`}
      ></ha-card>
    `;

  // return html`
  // <ha-card @click="${this._handleClick}">
  //   ${this._config.current !== false ? this.renderCurrent(stateObj) : ""}
  //   ${this._config.details !== false ? this.renderDetails(stateObj, lang) : ""}
  //   ${this._config.forecast !== false
  //     ? this.renderForecast(stateObj.attributes.forecast, lang)
  //     : ""}
  // </ha-card>
  // `;

    // if (!this._config || !this.hass) {
    //   return html``;
    // }

    // this.numberElements = 0;

    // const lang = this.hass.selectedLanguage || this.hass.language;
    // const stateObj = this.hass.states[this._config.entity];

    // if (!stateObj) {
    //   return html`
    //     <style>
    //       .not-found {
    //         flex: 1;
    //         background-color: yellow;
    //         padding: 8px;
    //       }
    //     </style>
    //     <ha-card>
    //       <div class="not-found">
    //         Entity not available: ${this._config.entity}
    //       </div>
    //     </ha-card>
    //   `;
    // }

  }

  private _handleAction(ev: ActionHandlerEvent): void {
    if (this.hass && this.config && ev.detail.action) {
      handleAction(this, this.hass, this.config, ev.detail.action);
    }
  }

  private _showWarning(warning: string): TemplateResult {
    return html` <hui-warning>${warning}</hui-warning> `;
  }

  private _showError(error: string): TemplateResult {
    const errorCard = document.createElement('hui-error-card');
    errorCard.setConfig({
      type: 'error',
      error,
      origConfig: this.config,
    });

    return html` ${errorCard} `;
  }


  // renderCurrent(stateObj) {
  //   this.numberElements++;

  //   return html`
  //     <div class="current ${this.numberElements > 1 ? "spacer" : ""}">
  //       <span
  //         class="icon bigger"
  //         style="background: none, url('${this.getWeatherIcon(
  //           stateObj.state.toLowerCase(),
  //           this.hass.states["sun.sun"]
  //         )}') no-repeat; background-size: contain;"
  //         >${stateObj.state}
  //       </span>
  //       ${this._config.name
  //         ? html` <span class="title"> ${this._config.name} </span> `
  //         : ""}
  //       <span class="temp"
  //         >${this.getUnit("temperature") == "°F"
  //           ? Math.round(stateObj.attributes.temperature)
  //           : stateObj.attributes.temperature}</span
  //       >
  //       <span class="tempc"> ${this.getUnit("temperature")}</span>
  //     </div>
  //   `;
  // }

  // renderDetails(stateObj, lang) {
  //   const sun = this.hass.states["sun.sun"];
  //   let next_rising;
  //   let next_setting;

  //   if (sun) {
  //     next_rising = new Date(sun.attributes.next_rising).toLocaleTimeString(lang, {
  //                       hour: "2-digit",
  //                       minute: "2-digit",
  //                   });
  //     next_setting = new Date(sun.attributes.next_setting).toLocaleTimeString(lang, {
  //                       hour: "2-digit",
  //                       minute: "2-digit",
  //                   });
  //   }

  //   this.numberElements++;

  //   return html`
  //     <ul class="variations ${this.numberElements > 1 ? "spacer" : ""}">
  //       <li>
  //         <ha-icon icon="mdi:water-percent"></ha-icon>
  //         ${stateObj.attributes.humidity}<span class="unit"> % </span>
  //       </li>
  //       <li>
  //         <ha-icon icon="mdi:weather-windy"></ha-icon> ${windDirections[
  //           parseInt((stateObj.attributes.wind_bearing + 11.25) / 22.5)
  //         ]}
  //         ${stateObj.attributes.wind_speed}<span class="unit">
  //           ${this.getUnit("length")}/h
  //         </span>
  //       </li>
  //       <li>
  //         <ha-icon icon="mdi:gauge"></ha-icon>
  //         ${stateObj.attributes.pressure}
  //         <span class="unit">
  //           ${this.getUnit("air_pressure")}
  //         </span>
  //       </li>
  //       <li>
  //         <ha-icon icon="mdi:weather-fog"></ha-icon> ${stateObj.attributes
  //           .visibility}<span class="unit">
  //           ${this.getUnit("length")}
  //         </span>
  //       </li>
  //       ${next_rising
  //         ? html`
  //             <li>
  //               <ha-icon icon="mdi:weather-sunset-up"></ha-icon>
  //               ${next_rising}
  //             </li>
  //           `
  //         : ""}
  //       ${next_setting
  //         ? html`
  //             <li>
  //               <ha-icon icon="mdi:weather-sunset-down"></ha-icon>
  //               ${next_setting}
  //             </li>
  //           `
  //         : ""}
  //     </ul>
  //   `;
  // }

  // renderForecast(forecast, lang) {
  //   if (!forecast || forecast.length === 0) {
  //     return html``;
  //   }

  //   this.numberElements++;
  //   return html`
  //     <div class="forecast clear ${this.numberElements > 1 ? "spacer" : ""}">
  //       ${forecast
  //         .slice(
  //           0,
  //           this._config.number_of_forecasts
  //             ? this._config.number_of_forecasts
  //             : 5
  //         )
  //         .map(
  //           (daily) => html`
  //             <div class="day">
  //               <div class="dayname">
  //                 ${this._config.hourly_forecast
  //                   ? new Date(daily.datetime).toLocaleTimeString(lang, {
  //                       hour: "2-digit",
  //                       minute: "2-digit",
  //                     })
  //                   : new Date(daily.datetime).toLocaleDateString(lang, {
  //                       weekday: "short",
  //                     })}
  //               </div>
  //               <i
  //                 class="icon"
  //                 style="background: none, url('${this.getWeatherIcon(
  //                   daily.condition.toLowerCase()
  //                 )}') no-repeat; background-size: contain"
  //               ></i>
  //               <div class="highTemp">
  //                 ${daily.temperature}${this.getUnit("temperature")}
  //               </div>
  //               ${daily.templow !== undefined
  //                 ? html`
  //                     <div class="lowTemp">
  //                       ${daily.templow}${this.getUnit("temperature")}
  //                     </div>
  //                   `
  //                 : ""}
  //               ${!this._config.hide_precipitation &&
  //               daily.precipitation !== undefined &&
  //               daily.precipitation !== null
  //                 ? html`
  //                     <div class="precipitation">
  //                       ${Math.round(daily.precipitation*10)/10} ${this.getUnit("precipitation")}
  //                     </div>
  //                   `
  //                 : ""}
  //               ${!this._config.hide_precipitation &&
  //               daily.precipitation_probability !== undefined &&
  //               daily.precipitation_probability !== null
  //                 ? html`
  //                     <div class="precipitation_probability">
  //                       ${Math.round(daily.precipitation_probability)} ${this.getUnit("precipitation_probability")}
  //                     </div>
  //                   `
  //                 : ""}
  //             </div>
  //           `
  //         )}
  //     </div>
  //   `;
  // }

  getWeatherIcon(condition, sun) {
    return `${
      this._config.icons
        ? this._config.icons
        : "https://cdn.jsdelivr.net/gh/bramkragten/weather-card/dist/icons/"
    }${
      sun && sun.state == "below_horizon"
        ? weatherIconsNight[condition]
        : weatherIconsDay[condition]
    }.svg`;
  }

  getUnit(measure) {
    const lengthUnit = this.hass.config.unit_system.length;
    switch (measure) {
      case "air_pressure":
        return lengthUnit === "km" ? "hPa" : "inHg";
      case "length":
        return lengthUnit;
      case "precipitation":
        return lengthUnit === "km" ? "mm" : "in";
      case "precipitation_probability":
        return "%";
      default:
        return this.hass.config.unit_system[measure] || "";
    }
  }

  // _handleClick() {
  //   fireEvent(this, "hass-more-info", { entityId: this._config.entity });
  // }

  getCardSize() {
    return 3;
  }

  // https://lit.dev/docs/components/styles/
  static get styles() {
    return css`
      ha-card {
        cursor: pointer;
        margin: auto;
        overflow: hidden;
        padding-top: 1.3em;
        padding-bottom: 1.3em;
        padding-left: 1em;
        padding-right: 1em;
        position: relative;
      }

      .spacer {
        padding-top: 1em;
      }

      .clear {
        clear: both;
      }

      .title {
        position: absolute;
        left: 3em;
        font-weight: 300;
        font-size: 3em;
        color: var(--primary-text-color);
      }

      .temp {
        font-weight: 300;
        font-size: 4em;
        color: var(--primary-text-color);
        position: absolute;
        right: 1em;
      }

      .tempc {
        font-weight: 300;
        font-size: 1.5em;
        vertical-align: super;
        color: var(--primary-text-color);
        position: absolute;
        right: 1em;
        margin-top: -14px;
        margin-right: 7px;
      }

      @media (max-width: 460px) {
        .title {
          font-size: 2.2em;
          left: 4em;
        }
        .temp {
          font-size: 3em;
        }
        .tempc {
          font-size: 1em;
        }
      }

      .current {
        padding: 1.2em 0;
        margin-bottom: 3.5em;
      }

      .variations {
        display: flex;
        flex-flow: row wrap;
        justify-content: space-between;
        font-weight: 300;
        color: var(--primary-text-color);
        list-style: none;
        padding: 0 1em;
        margin: 0;
      }

      .variations ha-icon {
        height: 22px;
        margin-right: 5px;
        color: var(--paper-item-icon-color);
      }

      .variations li {
        flex-basis: auto;
        width: 50%;
      }

      .variations li:nth-child(2n) {
        text-align: right;
      }

      .variations li:nth-child(2n) ha-icon {
        margin-right: 0;
        margin-left: 8px;
        float: right;
      }

      .unit {
        font-size: 0.8em;
      }

      .forecast {
        width: 100%;
        margin: 0 auto;
        display: flex;
      }

      .day {
        flex: 1;
        display: block;
        text-align: center;
        color: var(--primary-text-color);
        border-right: 0.1em solid #d9d9d9;
        line-height: 2;
        box-sizing: border-box;
      }

      .dayname {
        text-transform: uppercase;
      }

      .forecast .day:first-child {
        margin-left: 0;
      }

      .forecast .day:nth-last-child(1) {
        border-right: none;
        margin-right: 0;
      }

      .highTemp {
        font-weight: bold;
      }

      .lowTemp {
        color: var(--secondary-text-color);
      }

      .precipitation {
        color: var(--primary-text-color);
        font-weight: 300;
      }

      .icon.bigger {
        width: 10em;
        height: 10em;
        margin-top: -4em;
        position: absolute;
        left: 0em;
      }

      .icon {
        width: 50px;
        height: 50px;
        margin-right: 5px;
        display: inline-block;
        vertical-align: middle;
        background-size: contain;
        background-position: center center;
        background-repeat: no-repeat;
        text-indent: -9999px;
      }

      .weather {
        font-weight: 300;
        font-size: 1.5em;
        color: var(--primary-text-color);
        text-align: left;
        position: absolute;
        top: -0.5em;
        left: 6em;
        word-wrap: break-word;
        width: 30%;
      }
    `;
  }
}
// customElements.define("weather-card", VWeatherCard);
